feat: Jupyterhub with keycloak, spark and s3 (#155)
stackabletech/demos: docs/modules/demos/pages/jupyterhub-keycloak.adoc

The same users as in the xref:end-to-end-security.adoc[End-to-end-security] demo are configured in Keycloak and these will be used as examples.
The notebook offers a simple template for using Spark to interact with S3 as a storage backend.

== Overview

This demo will:

* Install the required Stackable Data Platform operators
* Spin up the following data products:
** *JupyterHub*: A multi-user server for Jupyter notebooks
** *Keycloak*: An identity and access management product
** *S3*: A Minio instance for data storage
* Download a sample of the gas sensor dataset into S3
* Install the Jupyter notebook
* Demonstrate some basic data operations against S3
* Illustrate multi-user usage

== JupyterHub

Have a look at the available Pods before logging in:

[source,console]
----
$ kubectl get pods
NAME                         READY   STATUS      RESTARTS   AGE
hub-84f49ccbd7-29h7j         1/1     Running     0          56m
keycloak-544d757f57-f55kr    2/2     Running     0          57m
load-gas-data-m6z5p          0/1     Completed   0          54m
minio-5486d7584f-x2jn8       1/1     Running     0          57m
proxy-648bf7f45b-62vqg       1/1     Running     0          56m

----

> The `proxy` Pod has an associated `proxy-public` service with a statically-defined port (31095), exposed with type NodePort.

Reviewer: ```suggestion
The `proxy` Pod has an associated `proxy-public` service with a statically-defined port (31095), exposed with type NodePort. The `keycloak` Pod has a Service called `keycloak` with a fixed port (31093) of type NodePort as well.
```

In order to reach the JupyterHub web interface, navigate to this service.
The node port IP can be found in the ConfigMap `keycloak-address` (written by the Keycloak Deployment as it starts up).
In the example below that would then be 172.19.0.5:31095:

[source,yaml]
----
apiVersion: v1
data:
  keycloakAddress: 172.19.0.5:31093 # Keycloak itself

>   keycloakNodeIp: 172.19.0.5 # can be used to access the proxy-public service

Reviewer: What if its on another node?

Author: I've just checked - you can reach the jupyterhub proxy-public from any node.

Author: This may be a KinD thing, though. Not sure.

Author: I'll add a comment.

Author: See 7be5288

image::jupyterhub-keycloak/server-options.png[]

The explorer window on the left includes a notebook that is already mounted.

Double-click on the file `notebook/process-s3.ipynb`:

image::jupyterhub-keycloak/load-nb.png[]

Run the notebook by selecting "Run All Cells" from the menu:

image::jupyterhub-keycloak/run-nb.png[]

The notebook includes some comments regarding image compatibility and uses a custom image built off the official Spark image that matches the Spark version used in the notebook.
The java versions also match exactly.
Python versions need to match at the `major:minor` level, which is why Python 3.11 is used in the custom image.

Once the spark executor has been started (we have specified `spark.executor.instances` = 1) it will spin up as an extra pod.
We have named the spark job to incorporate the current user (justin-martin).
JupyterHub has started a pod for the user's notebook instance (`jupyter-justin-martin---bdd3b4a1`) and another one for the spark executor (`process-s3-jupyter-justin-martin-bdd3b4a1-9e9da995473f481f-exec-1`):

[source,console]
----

> 12:49 $ kubectl get pods

Reviewer: ```suggestion
$ kubectl get pods
```

NAME                                   READY   STATUS      RESTARTS   AGE
...
jupyter-justin-martin---bdd3b4a1       1/1     Running     0          17m
process-s3-jupyter-justin-martin-...   1/1     Running     0          2m9s
...
----

Stop the kernel in the notebook (which will shut down the spark session and thus the executor) and log out as the current user.
Log in now as `daniel.king` and then again as `isla.williams` (you may need to do this in a clean browser sessions so that existing login cookies are removed).
This user has been defined as an admin user in the jupyterhub configuration:

[source,yaml]
----
  ...
  hub:
    config:
      Authenticator:
        # don't filter here: delegate to Keycloak
        allow_all: True
        admin_users:
          - isla.williams
  ...
----

You should now see user-specific pods for all three users:

[source,console]
----

> 16:16 $ kubectl get pods

Reviewer 2: ```suggestion
$ kubectl get pods
```

NAME                               READY   STATUS      RESTARTS   AGE
...
jupyter-daniel-king---181a80ce     1/1     Running     0          6m17s
jupyter-isla-williams---14730816   1/1     Running     0          4m50s
jupyter-justin-martin---bdd3b4a1   1/1     Running     0          3h47m
...
----

The admin user (`isla.williams`) will also have an extra Admin tab in the JupyterHub console where current users can be managed.
You can find this in the JupyterHub UI at http://<ip>:31095/hub/admin e.g http://172.19.0.5:31095/hub/admin:

image::jupyterhub-keycloak/admin-tab.png[]

You can inspect the S3 buckets by using stackable stacklet list to return the Minio endpoint and logging in there with `admin/adminadmin`:

[source,console]
----

> 15:15 $ stackablectl stacklet list

Reviewer 2: ```suggestion
$ stackablectl stacklet list
```

= jupyterhub-keycloak

:k8s-cpu: https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/#cpu
:spark-pkg: https://spark.apache.org/docs/latest/api/python/user_guide/python_packaging.html
:pyspark: https://spark.apache.org/docs/latest/api/python/getting_started/index.html
:jupyterhub-k8s: https://github.com/jupyterhub/zero-to-jupyterhub-k8s
:jupyterlab: https://jupyterlab.readthedocs.io/en/stable/
:jupyter: https://jupyter.org
:keycloak: https://www.keycloak.org/
:gas-sensor: https://archive.ics.uci.edu/dataset/487/gas+sensor+array+temperature+modulation

> This demo showcases the integration between {jupyter}[Jupyter] and {keycloak}[Keycloak] deployed on the Stackable Data Platform (SDP) Kubernetes cluster.

Reviewer: ```suggestion
This demo showcases the integration between {jupyter}[JupyterHub] and {keycloak}[Keycloak] deployed on the Stackable Data Platform (SDP) onto a Kubernetes cluster.
```

These Pods in turn can mount *all* volumes and Secrets in that namespace.
To prevent this from breaking user separation, it is planned to use an OPA gatekeeper to have OPA rules that restrict what the created executor Pods can mount. This is not yet implemented in this demo.
Secondly, this demo currently employs user impersonation for the spark job that writes to HDFS.
In a later iteration of this demo it is planned to replace this for HDFS with OPA group mapping for authorization and Kerberos for authentication.

[#system-requirements]
== System requirements

To run this demo, your system needs at least:

* 8 {k8s-cpu}[cpu units] (core/hyperthread)
* 32GiB memory

You may need more resources depending on how many concurrent users are logged in, and which notebook profiles they are using.

== Aim / Context

This demo shows how to authenticate JupyerHub users against a Keycloak backend using JupyterHub's OAuthenticator.
The same users as in the xref:end-to-end-security.adoc[End-to-end-security] demo are configured in Keycloak and these will be used as examples.

> The notebook offers a simple template for using spark to interact with different backends, specifically S3 and HDFS.

Reviewer: ```suggestion
The notebook offers a simple template for using Spark to interact with different backends, specifically S3 and HDFS.
```

* Demonstrate some basic data operations against S3 and HDFS
* Illustrate multi-user usage

== JupyterHub

Have a look at the available Pods before logging in:

[source,console]
----
$ kubectl get pods
NAME                         READY   STATUS      RESTARTS   AGE
create-hdfs-dir-jw7h4        0/1     Completed   0          54m
hdfs-datanode-default-0      1/1     Running     0          54m
hdfs-journalnode-default-0   1/1     Running     0          54m
hdfs-namenode-default-0      2/2     Running     0          54m
hdfs-namenode-default-1      2/2     Running     0          50m
hub-84f49ccbd7-29h7j         1/1     Running     0          56m
keycloak-544d757f57-f55kr    2/2     Running     0          57m
load-gas-data-m6z5p          0/1     Completed   0          54m
minio-5486d7584f-x2jn8       1/1     Running     0          57m
proxy-648bf7f45b-62vqg       1/1     Running     0          56m
zookeeper-server-default-0   1/1     Running     0          54m

----

The `proxy` Pod has an associated `proxy-public` service with a statically-defined port (31095), exposed with type NodePort.
In order to reach the JupyterHub web interface, navigate to this service.

> The node port IP can be found in the ConfigMap `keycloak-address` (written by the Keycloak deployment as it starts up).

Reviewer: ```suggestion
The node port IP can be found in the ConfigMap `keycloak-address` (written by the Keycloak Deployment as it starts up).
```

In the example below that would then be 172.19.0.5:31095:

[source,yaml]
----
apiVersion: v1
data:
  keycloakAddress: 172.19.0.5:31093 # Keycloak itself
  keycloakNodeIp: 172.19.0.5 # can be used to access the proxy-public service
kind: ConfigMap
metadata:
  name: keycloak-address
  namespace: default
----

> NOTE: The hub may show `CreateContainerConfigError`` for a few moments on start-up as it requires the ConfigMap written by the Keycloak deployment.

Reviewer: ```suggestion
NOTE: The `hub` Pod may show `CreateContainerConfigError` for a few moments on start-up as it requires the ConfigMap written by the Keycloak deployment.
```

You should see the JupyterHub login page, which will indicate a re-direct to the OAuth service (Keycloak):

image::jupyterhub-keycloak/oauth-login.png[]

Click on the sign-in button.

> You will redirected to the Keycloak login, where you can enter one of the afore-mentioned users:

Reviewer: Dictionary says `aforementioned`?
```suggestion
You will be redirected to the Keycloak login, where you can enter one of the aforementioned users:
```

> You will redirected to the Keycloak login, where you can enter one of the afore-mentioned users:

Author: Done: 80bd2c6

image::jupyterhub-keycloak/keycloak-login.png[]

A successful login will redirect you back to JupyterHub where different profiles are listed (the drop-down options are visible when you click on the respective fields):

image::jupyterhub-keycloak/server-options.png[]

The explorer window on the left includes HDFS configuration files and a notebook that is already mounted.
Open the `hdfs` folder where the `core-site.xml` and `hdfs-site.xml` from the discovery ConfigMap of the HDFS cluster are located.
You can see how to reference the HDFS cluster when reading from it in the notebook (`hdfs://hdfs/`):

image::jupyterhub-keycloak/hdfs-core-site.png[]

Double-click on the file `notebook/process-s3.ipynb`:

image::jupyterhub-keycloak/load-nb.png[]

Run the notebook by selecting "Run All Cells" from the menu:

image::jupyterhub-keycloak/run-nb.png[]

> The notebook includes some comments regarding image compatibility and uses a custom image built off the official Spark official that matches the Spark version used in the notebook.

Reviewer: ```suggestion
The notebook includes some comments regarding image compatibility and uses a custom image built off the official Spark image that matches the Spark version used in the notebook.
```

The java versions also match exactly.
Python versions need to match at the `major:minor` level, which is why Python 3.11 is used in the custom image.

Once the spark executor has been started (we have specified `spark.executor.instances` = 1) it will spin up as an extra pod.
We have named the spark job to incorporate the current user (justin-martin).
JupyterHub has started a pod for the user's notebook instance (`jupyter-justin-martin---bdd3b4a1`) and another one for the spark executor (`process-s3-jupyter-justin-martin-bdd3b4a1-9e9da995473f481f-exec-1`):

[source,console]
----
12:49 $ kubectl get pods
NAME                                   READY   STATUS      RESTARTS   AGE
...
jupyter-justin-martin---bdd3b4a1       1/1     Running     0          17m
process-s3-jupyter-justin-martin-...   1/1     Running     0          2m9s
...
----

Stop the kernel in the notebook (which will shut down the spark session and thus the executor) and log out as the current user.

> Log in now as `daniel.king` and then again `isla.williams` (you may need so do this in a clean browser sessions so that existing login cookies are removed).

Reviewer: ```suggestion
Log in now as `daniel.king` and then again as `isla.williams` (you may need to do this in a clean browser sessions so that existing login cookies are removed).
```

This user has been defined as an admin user in the jupyterhub configuration:

[source,yaml]
----
  ...
  hub:
    config:
      Authenticator:
        # don't filter here: delegate to Keycloak
        allow_all: True
        admin_users:
          - isla.williams
  ...
----

You should now see user-specific pods for all three users:

[source,console]
----
16:16 $ kubectl get pods
NAME                               READY   STATUS      RESTARTS   AGE
...
jupyter-daniel-king---181a80ce     1/1     Running     0          6m17s
jupyter-isla-williams---14730816   1/1     Running     0          4m50s
jupyter-justin-martin---bdd3b4a1   1/1     Running     0          3h47m
...
----

> The admin user (`isla.williams`) will also have an extra Admin tab in the JupyterHub console where current users can be managed:

Reviewer: Needed to look for that tab, could we add the ref `/hub/admin#/` or `http://172.18.0.2:31095/hub/admin#/` here?

Author: Done: 80bd2c6
